### PR TITLE
ParseDate.parse_range method implemented

### LINKED ISSUE #10: add class method ParseDate.parse_range(str)
~~blocked by #9 ~~

Add a class method `parse_range` to ParseDate that, given a string, returns the range of years as an array. Utilize the new `range_array` method.

Basically, given string, parse earliest year, latest year, and then get the range with those endpoints.

Inspired by the guts of this DLME macro:

https://github.com/sul-dlss/dlme-transform/blob/master/lib/macros/date_parsing.rb#L28-L42

### PR DESCRIPTION
**Pull request by ndushay**
## Why was this change made?

Closes #10

`parse_range` is intended for parsing out a range of years from a date string, generally intended for a date slider in a UI.  This functionality is very useful to DLME, and will greatly simplify the macros needed in the dlme-transform code base.

It makes more sense for this logic to live in this gem as it is potentially useful for date parsing beyond DLME.

Note that after this is merged, DLME date parsing willl be refactored to use `DateParse.parse_range` when appropriate.  (see sul-dlss/dlme-transform/issues/291)

## Was the documentation (README, API, wiki, ...) updated?

yes, README

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,91 +30,127 @@ ParseDate has class methods for date string parsing.
 ```
 require 'parse_date'
 
-ParseDate.earliest_year('12/25/00') # 2000
```
Comment (ndushay) [inline README.md]: the only change i made to the existing lines was to put the expected values tabbed to the right

File: lib/parse_date/int_from_string.rb [hunk 2]
```diff
@@ -94,17 +94,12 @@ def remove_brackets(date_str)
       date_str.delete('[]') if date_str.match(BRACKETS_BETWEEN_DIGITS_REGEX)
     end
 
-    YYYY_HYPHEN_YYYY_REGEX = Regexp.new(/(?<first>\d{4})\??\s*-\s*(?<last>\d{1,4})\??/m)
+    YYYY_HYPHEN_YYYY_REGEX = Regexp.new(/(?<first>\d{4})\??\s*-\s*(?<last>\d{4})\??/m)
```
Comment (ndushay) [inline lib/parse_date/int_from_string.rb]: this small regex tweak let me simplify the method `hyphen_4digit_latest_year`

File: lib/parse_date/int_from_string.rb [hunk 3]
```diff
@@ -127,8 +122,7 @@ def hyphen_2digit_latest_year(date_str)
     # Integer value for latest year if we have "yyuu-yyuu" pattern
     # @return [Integer, nil] yyyy if date_str matches pattern; nil otherwise
     def yyuu_after_hyphen(date_str)
-      matches = date_str.match(YYuu_HYPHEN_YYuu_REGEX)
-      last_year_for_century(Regexp.last_match(:last)).to_i if matches
+      last_year_for_century(Regexp.last_match(:last)).to_i if date_str.match(YYuu_HYPHEN_YYuu_REGEX)
```
Comment (ndushay) [inline lib/parse_date/int_from_string.rb]: I have no idea why I didn't use this simpler expression in the first place ...

File: lib/parse_date/int_from_string.rb [hunk 4]
```diff
@@ -255,49 +247,55 @@ def last_year_for_bc_century(date_str)
...
     end
 
-    EARLY_NUMERIC_REGEX = Regexp.new('^\-?\d{1,3}$', REGEX_OPTS)
+    EARLY_NUMERIC_REGEX = Regexp.new('^\-?\d{1,3}([^\du\-\[]|$)', REGEX_OPTS)
 
     # year if date_str contains yyy, yy, y, -y, -yy, -yyy, -yyyy
     # @return [Integer, nil] year if date_str matches pattern; nil otherwise
     def year_for_early_numeric(date_str)
-      date_str.to_i if date_str.match(EARLY_NUMERIC_REGEX) || date_str =~ /^-\d{4}$/
+      date_str.to_i if date_str.match(EARLY_NUMERIC_REGEX) || date_str =~ /^-\d{4}([^\du\-\[]|$)$/
```
Comment (ndushay) [inline lib/parse_date/int_from_string.rb]: this weeds out 5 digit numbers and other variants

File: lib/parse_date/int_from_string.rb [hunk 5]
```diff
@@ -255,49 +247,55 @@ def last_year_for_bc_century(date_str)
...
+
+    FIRST_LAST_EARLY_NUMERIC_REGEX =
+      Regexp.new(/^(?<first>\-?\d{1,3})\??\s*(-|–|or)\s*(?<last>\-?\d{1,4})\??([^\du\-\[]|$)/im)
+
+    # Integer value for latest year if we have early numeric year range or single early numeric year
+    # @return [Integer, nil] year if date_str matches pattern; nil otherwise
+    def last_year_for_early_numeric(date_str)
+      return Regexp.last_match(:last).to_i if date_str.match(FIRST_LAST_EARLY_NUMERIC_REGEX)
+
+      year_for_early_numeric(date_str) # if single year, not matched above
```
Comment (ndushay) [inline lib/parse_date/int_from_string.rb]: whee!   let's allow hyphenated expression for early years!

File: README.md [hunk 6]
```diff
@@ -30,91 +30,127 @@ ParseDate has class methods for date string parsing.
...
+ParseDate.latest_year('1757-58')                        # 1758
+ParseDate.latest_year('1975-05')                        # 1975 (range invalid)
+ParseDate.latest_year('1835 or 1836')                   # 1836
+ParseDate.latest_year('17-- or 18--?')                  # 1899
+ParseDate.latest_year('17th or 18th century?')          # 1799
+ParseDate.latest_year('ca. 5th–6th century A.D.')       # 599
+ParseDate.latest_year('ca. 9th–8th century B.C.')       # -800
+ParseDate.latest_year('ca. 13th–12th century B.C.')     # -1200
+ParseDate.latest_year('5th century B.C.')               # -500
+ParseDate.parse_range('-5 - 3')                         # 3,
```
Comment (jmartin-sul) [inline README.md]: should this `parse_range` be in the section above w/ the other `parse_range` examples?  or actually, should it be a `latest_year` example instead?

File: README.md [hunk 7]
```diff
@@ -30,91 +30,127 @@ ParseDate has class methods for date string parsing.
...
+ParseDate.latest_year('18th century CE')                # 1799
+ParseDate.latest_year('17uu')                           # 1799
+ParseDate.latest_year('between 1694 and 1799')          # 1799
+ParseDate.latest_year('between 1 and 5')                # 5
+ParseDate.latest_year('between 300 and 150 B.C.')       # -150
+ParseDate.latest_year('1496-1499')                      # 1499
+ParseDate.latest_year('1750?-1867')                     # 1867
+ParseDate.latest_year('17--?-18--?')                    # 1899
+ParseDate.latest_year('1757-58')                        # 1758
+ParseDate.latest_year('1975-05')                        # 1975 (range invalid)
```
Comment (jmartin-sul) [inline README.md]: i know you were saying you just aligned cols on these, but does this raise an error?  if so, worth noting?

File: spec/parse_date_spec.rb [hunk 8]
```diff
@@ -5,6 +5,79 @@
...
+        '18th century CE' => (1700..1799).to_a,
+        '17uu' => (1700..1799).to_a,
+        'between 1694 and 1799' => (1694..1799).to_a,
+        'between 1 and 5' => (1..5).to_a,
+        'between 300 and 150 B.C.' => (-300..-150).to_a,
+        '-5 - 3' => (-5..3).to_a,
+        '1496-1499' => (1496..1499).to_a,
+        '1750?-1867' => (1750..1867).to_a,
+        '17--?-18--?' => (1700..1899).to_a,
+        '1835 or 1836' => [1835, 1836].to_a,
```
Comment (jmartin-sul) [inline spec/parse_date_spec.rb]: `.to_a` seems superfluous here, but def wouldn't hold up merge over that

File: README.md (hunk 7 above)
Comment (ndushay) [reply README.md]: no, in this case, yyyy-yy, it is sometimes a month, so it makes other assumptions.  2001-05 would be interpreted as 2005 for latest year, but 2019-05 would be 2019.

File: README.md (hunk 6 above)
Comment (ndushay) [reply README.md]: typo - will fix, thanks.  Copy pasta -- I got it recognizing the 3 for latest year.

File: README.md (hunk 7 above)
Comment (jmartin-sul) [reply README.md]: ah, got it, ok.  thanks.

File: spec/parse_date_spec.rb (hunk 8 above)
Comment (ndushay) [reply spec/parse_date_spec.rb]: good 👀 - thanks.

**Pull request merged**